### DISCUSSION TITLE
How i can construct object from audit data?

### DISCUSSION THREAD

**Discussion by faketogosamogo** [thepirat000/Audit.NET#411] [Q&A]
Hellow. Hello, is it possible to construct an object at a certain point in time based on audit data? I looked for it in your documentation, but I didn't find it. If this library does not have such a feature, can you tell me how to do it.

**Comment by thepirat000** [comment]
I guess you are referring to querying past audit events that you saved with a data provider.

It depends on the data provider, some of them provides `GetEvent()` / `QueryEvents()` methods for example the [AzureCosmos data provider](https://github.com/thepirat000/Audit.NET/blob/master/src/Audit.NET.AzureCosmos/README.md#query-events)

**Comment by faketogosamogo** [comment]
No, for example: i have: 
```
"StartDate": "2021-05-28T08:03:02.6144793Z",
  "EndDate": "2021-05-28T08:03:03.8741165Z",
  "Duration": 1260,
  "EntityFrameworkEvent": {
    "Database": "AuditCheck",
    "Entries": [
      {
        "Table": "Forms",
        "Name": "Form",
        "Action": "Insert",
        "PrimaryKey": {
          "Id": "8d9a47e8-645a-4683-8d05-1de082242c2b"
        },
        "Entity": {
          "Id": "8d9a47e8-645a-4683-8d05-1de082242c2b",
          "Name": "123"
        },
        "ColumnValues": {
          "Id": "8d9a47e8-645a-4683-8d05-1de082242c2b",
          "Name": "123"
        },
        "Valid": true
      }
    ],
    "Result": 1,
    "Success": true
  }
  "StartDate": "2021-05-28T08:48:28.1874384Z",
  "EndDate": "2021-05-28T08:48:28.3056158Z",
  "Duration": 118,
  "EntityFrameworkEvent": {
    "Database": "AuditCheck",
    "Entries": [
      {
        "Table": "Forms",
        "Name": "Form",
        "Action": "Update",
        "PrimaryKey": {
          "Id": "8d9a47e8-645a-4683-8d05-1de082242c2b"
        },
        "Changes": [
          {
            "ColumnName": "Name",
            "OriginalValue": "123",
            "NewValue": "fas"
          }
        ],
        "ColumnValues": {
          "Id": "8d9a47e8-645a-4683-8d05-1de082242c2b",
          "Name": "fas"
        },
        "Valid": true
      }
    ],
    "Result": 1,
    "Success": true
  }
```

How can I construct my Entity from this data at the time I need?

**Comment by thepirat000** [comment]
Where do you have the JSON you shared, in a file?

And what is _your Entity_, a .NET object?

Please share the context otherwise I cannot help

**Comment by faketogosamogo** [comment]
I connected Audit.Net Entity Framework and this Files automaticaly created and I had records in database with audit data. 
```
 "EntityFrameworkEvent": {
    "Database": "AuditCheck",
    "Entries": [
      {
        "Table": "Forms",
        "Name": "Form",
        "Action": "Update",
        "PrimaryKey": {
          "Id": "8d9a47e8-645a-4683-8d05-1de082242c2b"
        }, 
```
can i get from this files or database data my Entity "FORM" from table "FORMS" at a certain point in time?

**Comment by faketogosamogo** [comment]
for example: 
10:00 i changed name form with Id: 1
11:00 i changed name form with Id: 1
12:00 i changed name form with Id: 1

and I need to get what the form looked like at 11:00

**Comment by thepirat000** [comment]
I'm not sure if you can construct a query like that. You can't really know the history of an entity just by looking at the client changes. Unless every DB action is audited, and even then it will be difficult to know, since you must look back for the latest change of every column, etc.

But that really depends on the [data provider](https://github.com/thepirat000/Audit.NET#data-providers) you use, or how you store the audit event info. 

I guess you didn't configure any data provider and it's using the default which is a [FileDataProvider](https://github.com/thepirat000/Audit.NET/blob/master/src/Audit.NET/Providers/FileDataProvider.cs) that stores the events as JSON files on the current directory.

I don't think storing the events in files will help you with that task.

If you need to do such queries you should probably use the [SqlServerDataProvider](https://github.com/thepirat000/Audit.NET/tree/master/src/Audit.NET.SqlServer#auditnetsqlserver) or the [EntityFramerowkDataProvider](https://github.com/thepirat000/Audit.NET/blob/master/src/Audit.EntityFramework/README.md#entity-framework-data-provider) to store the data in a database that you can query.

**Comment by thepirat000** [comment]
Also check the [IncludeEntityObjects setting](https://github.com/thepirat000/Audit.NET/blob/master/src/Audit.EntityFramework/README.md#settings), that could help since it will store the complete entity graph that is currently being tracked
